modules/metastorage/src/main/java/org/apache/ignite/metastorage/internal/watch/WatchAggregator.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.metastorage.internal.watch;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiConsumer;
+import org.apache.ignite.lang.IgniteBiTuple;
+import org.apache.ignite.metastorage.common.Key;
+import org.apache.ignite.metastorage.common.WatchEvent;
+import org.apache.ignite.metastorage.common.WatchListener;
+import org.jetbrains.annotations.NotNull;
+
+/** */
+public class WatchAggregator {
+    /**
+     * Watches' map must be synchronized because of changes from WatchListener in separate thread.
+     */
+    private final Map<Long, Watch> watches = Collections.synchronizedMap(new LinkedHashMap<>());
+
+    /** */
+    private long idCntr;
```

Comment:
As long as `MetaStorageManager#register...` methods are synchronized, it looks like a safe solution.
But, it would be better to use `AtomicLong` instead.
What do you think?